File: leku/src/main/java/com/schibstedspain/leku/geocoder/api/AddressBuilder.java
package com.schibstedspain.leku.geocoder.api;

import android.location.Address;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class AddressBuilder {

  public List<Address> parseResult(String json) throws JSONException {
    List<Address> addresses = new ArrayList<>();
    JSONObject root = new JSONObject(json);
    JSONArray results = root.getJSONArray("results");
    for (int i = 0; i < results.length(); i++) {
      addresses.add(parseAddress(results.getJSONObject(i)));
    }
    return addresses;
  }

  private Address parseAddress(JSONObject jsonObject) throws JSONException {
    JSONObject location = jsonObject.getJSONObject("geometry").getJSONObject("location");
    double latitude = location.getDouble("lat");
    double longitude = location.getDouble("lng");

    List<AddressComponent> components = getAddressComponents(jsonObject.getJSONArray("address_components"));


    String postalCode = "";
    String city = "";
    String number = "";
    String street = "";
    for (AddressComponent component : components) {
      if (component.types.contains("postal_code")) {
        postalCode = component.name;
      }
      if (component.types.contains("locality")) {
        city = component.name;
      }
      if (component.types.contains("street_number")) {
        number = component.name;
      }
      if (component.types.contains("route")) {
        street = component.name;
      }
    }
    StringBuilder fullAddress = new StringBuilder();
    fullAddress.append(street);
    if (!street.isEmpty() && !number.isEmpty()) {
      fullAddress.append(", ").append(number);
    }
    Address address = new Address(Locale.getDefault());
    address.setLatitude(latitude);
    address.setLongitude(longitude);
    address.setPostalCode(postalCode);
    address.setAddressLine(0, fullAddress.toString());
    address.setAddressLine(1, postalCode);
    address.setAddressLine(2, city);
    address.setLocality(city);
    return address;
  }

  private List<AddressComponent> getAddressComponents(JSONArray jsonComponents) throws JSONException {
    List<AddressComponent> components = new ArrayList<>();
    for (int i = 0; i < jsonComponents.length(); i++) {
      AddressComponent component = new AddressComponent();
      JSONObject jsonComponent = jsonComponents.getJSONObject(i);
      component.name = jsonComponent.getString("long_name");
      component.types = new ArrayList<>();
      JSONArray jsonTypes = jsonComponent.getJSONArray("types");
      for (int j = 0; j < jsonTypes.length(); j++) {
        component.types.add(jsonTypes.getString(j));
      }
      components.add(component);
    }
    return components;
  }

  private static class AddressComponent {
    String name;
    List<String> types;
  }
}

File: leku/src/main/java/com/schibstedspain/leku/geocoder/api/AddressBuilder.kt
package com.schibstedspain.leku.geocoder.api

import android.location.Address
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.Locale

class AddressBuilder {

    @Throws(JSONException::class)
    fun parseResult(json: String): List<Address> {
        val addresses = ArrayList<Address>()
        val root = JSONObject(json)
        val results = root.getJSONArray("results")
        for (i in 0 until results.length()) {
            addresses.add(parseAddress(results.getJSONObject(i)))
        }
        return addresses
    }

    @Throws(JSONException::class)
    private fun parseAddress(jsonObject: JSONObject): Address {
        val location = jsonObject.getJSONObject("geometry").getJSONObject("location")
        val latitude = location.getDouble("lat")
        val longitude = location.getDouble("lng")
        val components = getAddressComponents(jsonObject.getJSONArray("address_components"))
        var postalCode: String? = ""
        var city: String? = ""
        var number: String? = ""
        var street: String? = ""

        for (component in components) {
            if (component.types!!.contains("postal_code")) {
                postalCode = component.name
            }
            if (component.types!!.contains("locality")) {
                city = component.name
            }
            if (component.types!!.contains("street_number")) {
                number = component.name
            }
            if (component.types!!.contains("route")) {
                street = component.name
            }
        }

        val fullAddress = StringBuilder()
        fullAddress.append(street)
        if (!street!!.isEmpty() && !number!!.isEmpty()) {
            fullAddress.append(", ").append(number)
        }

        val address = Address(Locale.getDefault())
        address.latitude = latitude
        address.longitude = longitude
        address.postalCode = postalCode
        address.setAddressLine(0, fullAddress.toString())
        address.setAddressLine(1, postalCode)
        address.setAddressLine(2, city)
        address.locality = city
        return address
    }

    @Throws(JSONException::class)
    private fun getAddressComponents(jsonComponents: JSONArray): List<AddressComponent> {
        val components = ArrayList<AddressComponent>()
        for (i in 0 until jsonComponents.length()) {
            val component = AddressComponent()
            val jsonComponent = jsonComponents.getJSONObject(i)
            component.name = jsonComponent.getString("long_name")
            component.types = ArrayList()
            val jsonTypes = jsonComponent.getJSONArray("types")
            for (j in 0 until jsonTypes.length()) {
                component.types!!.add(jsonTypes.getString(j))
            }
            components.add(component)
        }
        return components
    }

    private class AddressComponent {
        internal var name: String? = null
        internal var types: MutableList<String>? = null
    }
}
